jupyter_notebook/analytics.py

```diff
@@ -0,0 +1,194 @@
+
+    def change_consent_status(self, status):
+        ci = self.control_info()
+        if status == CONSENT_OK:
+            if not ci.get('consent', False):
+                self.enable()
+            ci['consent'] = True
+        elif status == CONSENT_DENIED:
+            ci['consent'] = False
+        else:
+            ci.pop('consent', None)
+
+        self.store_control_info()
+
+    def last_submission(self):
+        ci = self.control_info()
+        if 'last_submission' in ci:
+            return parse_date(ci['last_submission'])
+
+        return None
+
+    def startup(self):
+        webapp = self.nbapp.web_app
+
+        consent = self.consent_status()
+        if consent == CONSENT_OK:
+            self.enable()
+        elif consent == CONSENT_UNKNOWN:
+            webapp.settings['analytics_prompt'] = True
+
+        base_url = webapp.settings['base_url']
+        webapp.add_handlers(".*$", [
+            (ujoin(base_url, r"/analytics/consent"), ConsentHandler,
+                {'sender': self}),
+        ])
+
+    def enable(self):
+        self.enabled = True
+        self.nbapp.log.info("Enabling analytics")
+        self.read_data()
+        self.record_version('jupyter_notebook', __version__)
+        self.record_version('server_python', sys.version)
+        self.schedule_send()
+
+        self.socket = zmq.Context.instance().socket(zmq.PULL)
+        port = self.socket.bind_to_random_port('tcp://127.0.0.1')
+        self.nbapp.kernel_manager.analytics_port = port
+        self.stream = ZMQStream(self.socket)
+        self.stream.on_recv(self.on_recv)
+
+    def on_recv(self, msg_parts):
+        data = json.loads(msg_parts[0])
+        msg_type = data['msg_type']
+        if msg_type == 'version':
+            self.record_version(data['project'], data['version'])
+        elif msg_type == 'feature':
+            self.record_feature(data['project'], data['feature'])
+
+    def schedule_send(self):
+        last = self.last_submission()
+        if last is None:
+            self.ioloop.add_callback(self.send)
+            return
+
+        since_last = datetime.utcnow() - last
+        if since_last > timedelta(days=28):
+            self.ioloop.add_callback(self.send)
+        else:
+            to_next = (last + timedelta(days=28)) - datetime.utcnow()
+            self.ioloop.call_later(to_next.total_seconds(), self.send)
+
+    def send(self):
+        http_client = AsyncHTTPClient()
+        # http_client.fetch('https://...', method='POST',
+        #                   body=json.dumps(self.data))
+        self.control_info()['last_submission'] = datetime.utcnow().isoformat()
+        self.store_control_info()
+
+        # Reset the data for the next month's collection
+        self.data = {}
+        self.record_version('jupyter_notebook', __version__)
+        self.record_version('server_python', sys.version)
+        self.ioloop.call_later(timedelta(days=28).total_seconds(), self.send)
```

Comment:
Would the ioloop have to be running an entire 28 days to run this? I certainly don't keep a notebook server running that long.